pkg/receive/hashring.go

```diff
@@ -241,13 +241,42 @@ func (c ketamaHashring) GetN(tenant string, ts *prompb.TimeSeries, n uint64) (En
 	return c.endpoints[endpointIndex], nil
 }
 
+type tenantSet map[string]tenantMatcher
+
+func (t tenantSet) match(tenant string) (bool, error) {
+	// Fast path for the common case of direct match.
+	if mt, ok := t[tenant]; ok && isExactMatcher(mt) {
+		return true, nil
+	} else {
+		for tenantPattern, matcherType := range t {
+			switch matcherType {
+			case TenantMatcherGlob:
+				matches, err := filepath.Match(tenantPattern, tenant)
+				if err != nil {
+					return false, fmt.Errorf("error matching tenant pattern %s (tenant %s): %w", tenantPattern, tenant, err)
+				}
+				if matches {
```

Comment:
should short circuit when a match is found, old code will continue and ignore previous matched pattern